[exo-deploy 6] Extract into two commands

hugoesthere opened the pull request:
Extracts `exo-deploy` into two commands: one for initializing deployment and the other for running actual deployment. Also adds step that actually runs `terraform apply`

Commit:
extract app deployer class

Commit:
Merge branch 'master' into hh-exo-deploy1

Commit:
initial commit

Commit:
add comments

Commit:
fix tests

Commit:
Merge branch 'master' into hh-exo-deploy1

Commit:
Merge branch 'hh-exo-deploy1' into hh-exo-deploy2

Commit:
resolve merge conflicts

Commit:
some work

Commit:
address comments and some more updates

Commit:
resolve merge conflicts

Commit:
fix whitespace

Commit:
remove terraformDir field

Commit:
resolve merge conflicts

Commit:
plan works

Commit:
some cleanup

Commit:
handle secrets

Commit:
resolve merge conflicts

Commit:
extract constructor

Commit:
address comment

Commit:
initial commit

Commit:
add aws-sdk-go to vendor

Commit:
configure aws

Commit:
Merge branch 'master' into hh-exo-deploy2

Commit:
resolve merge conflicts

Commit:
fix process helpers

Commit:
Merge branch 'hh-exo-deploy2' into hh-exo-deploy3

Commit:
remove exo-setup

Commit:
resolve merge conflicts

Commit:
Revert "resolve merge conflicts"

This reverts commit 4f0522cc5af3ff14f96dbd12b8a7a63cfba1f362.

Commit:
Merge branch 'hh-exo-deploy2' into hh-exo-deploy3

Commit:
remove submodule

Commit:
glide up

Commit:
extract aws/terraform helper

Commit:
add comment

Commit:
env if statement

Commit:
Merge branch 'hh-exo-deploy2' into hh-exo-deploy3

Commit:
init account

Commit:
command line version

Commit:
create secrets

Commit:
basic cr

Commit:
robustness

Commit:
Merge branch 'master' into hh-exo-deploy2

Commit:
resolve merge conflicts

Commit:
Merge branch 'hh-exo-deploy3' into hh-exo-deploy4

Commit:
linter fixes

Commit:
remove glide lock

Commit:
Merge branch 'hh-exo-deploy3' into hh-exo-deploy4

Commit:
address comments before writing tests

Commit:
write tests

Commit:
fix linter errors

Commit:
add test

Commit:
fix configure read

Commit:
cleanup

Commit:
cleanup

Commit:
cleanup

Commit:
cleanup

Commit:
remove tf string

Commit:
resolve merge conflicts

Commit:
Merge branch 'master' into hh-exo-deploy2

Commit:
resolve merge conflicts

Commit:
move into terraform dir

Commit:
resolve merge conflicts

Commit:
Merge branch 'hh-exo-deploy3' into hh-exo-deploy4

Commit:
remove app deployer

Commit:
address comments

Commit:
fix unmerged file

Commit:
Merge branch 'hh-exo-deploy3' into hh-exo-deploy4

Commit:
uncomment lint block

Commit:
Merge branch 'hh-exo-deploy3' into hh-exo-deploy4

Commit:
update print comments

Commit:
update print comments

Commit:
change create to post/put

Commit:
Revert "change create to post/put"

This reverts commit 2805ebc816b8c3f8e11b73fdde243c9ff029071c.

Commit:
fix hasBucket

Commit:
fix command options

Commit:
fix cobra description

Commit:
move conflic check to prompt

Commit:
move conflict check to prompt

Commit:
Merge branch 'master' into hh-exo-deploy2

Commit:
Merge branch 'hh-exo-deploy2' into hh-exo-deploy3

Commit:
Merge branch 'hh-exo-deploy3' into hh-exo-deploy4

Commit:
Merge branch 'hh-exo-deploy4' into hh-exo-deploy5

Commit:
cleanup

Commit:
checkout master

Commit:
resolve merge conflicts

Commit:
rename aws

Commit:
rename aws

Commit:
glide work

Commit:
Merge branch 'hh-exo-deploy3' into hh-exo-deploy4

Commit:
Revert "glide work"

This reverts commit a3cb50654d060597c189ff9e2207964c04f60869.

Commit:
glide work

Commit:
Merge branch 'hh-exo-deploy3' into hh-exo-deploy4

Commit:
Merge branch 'hh-exo-deploy4' into hh-exo-deploy5

Commit:
rename aws

Commit:
Merge branch 'hh-exo-deploy4' into hh-exo-deploy5

Commit:
chagnes

Commit:
simplify

Commit:
Merge branch 'hh-exo-deploy4' into hh-exo-deploy5

Commit:
swap secrets

Commit:
Merge branch 'hh-exo-deploy4' into hh-exo-deploy5

Commit:
resolve merge conflicts

Commit:
write delete

Commit:
update descriptions

Commit:
fix tests

Commit:
Merge branch 'hh-exo-deploy4' into hh-exo-deploy5

Commit:
fix tests

Commit:
fix tests

Commit:
Merge branch 'hh-exo-deploy4' into hh-exo-deploy5

Commit:
resolve merge conflicts

Commit:
Merge branch 'hh-exo-deploy4' into hh-exo-deploy5

Commit:
address comments

Commit:
Merge branch 'hh-exo-deploy4' into hh-exo-deploy5

Commit:
consolidate prompts

Commit:
fix tests

Commit:
Merge branch 'hh-exo-deploy4' into hh-exo-deploy5

Commit:
linters pass

Commit:
linters pass

Commit:
Merge branch 'master' into hh-exo-deploy4

Commit:
Merge branch 'hh-exo-deploy4' into hh-exo-deploy5

Commit:
remove test

Commit:
resolve merge conflicts

Commit:
refactor

Commit:
refactor

Commit:
refactorf

Commit:
refactor

Commit:
Merge branch 'master' into hh-exo-deploy4

Commit:
Merge branch 'hh-exo-deploy4' into hh-exo-deploy5

Commit:
remove cli

Commit:
remove more dirs

Commit:
Merge branch 'hh-exo-deploy5' into hh-exo-deploy6

Commit:
refactor

Commit:
fix linter

Commit:
secrets integration

Commit:
resolve merge conflicts

Commit:
resolve merge conflicts

Commit:
add prints

Commit:
resolve merge conflicts

Commit:
add print

Commit:
resolve merge conflicts

Commit:
fix merge

Commit:
image names

Commit:
Merge branch 'master' into hh-refactor

Commit:
Merge branch 'hh-refactor' into hh-read-secrets

Commit:
Merge branch 'hh-read-secrets' into hh-exo-deploy6

Commit:
fix tests

Commit:
address comments

Commit:
resolve merge conflicts

Commit:
update dirs

Commit:
resolve merge conflicts

Commit:
resolve merge conflicts

Commit:
change logger

Commit:
resolve merge conflicts

Commit:
resolve merge conflicts

Commit:
add print

Commit:
add service routes

Commit:
resolve merge conflicts

Commit:
Merge branch 'hh-exo-deploy6' into hh-exocom-routes

Commit:
source git urls

Commit:
fix tests

Commit:
Merge branch 'master' into hh-exo-deploy6

Commit:
Merge branch 'hh-exo-deploy6' into hh-exocom-routes

Commit:
Merge branch 'hh-service-env-vars' into hh-terraform-link

Commit:
initial spike

Commit:
Merge branch 'master' into hh-exo-deploy6

Commit:
Merge branch 'hh-exo-deploy6' into hh-exocom-routes

Commit:
Merge branch 'hh-service-env-vars' into hh-terraform-link

Commit:
Merge branch 'hh-terraform-link' into hh-exo-deploy

Commit:
update comments

Commit:
address comments

Commit:
add tests

Commit:
remove todo

Commit:
add commit hash var

Commit:
change var name

kevgo commented on src/cmd/deploy.go:
```diff
@@ -4,57 +4,45 @@ import (
-	Use:   "deploy",
-	Short: "Deploys Exosphere application to the cloud",
-	Long:  "Deploys Exosphere application to the cloud",
+var deployInit = &cobra.Command{
```
Extract this into a file `deploy_init.go`?

hugoesthere replied:
We've been keeping subcommands in the same file as the root command (see `cmd/configure.go`). This is a subcommand of `exo deploy` invoked by `exo deploy init`.

Commit:
edit

kevgo commented on src/cmd/deploy.go:
```diff
@@ -4,57 +4,45 @@ import (
+}
+
+var deployCmd = &cobra.Command{
+	Use:   "deploy",
```
How is this command called? `exo deploy` or `exo deploy deploy`?

hugoesthere replied:
Just `exo deploy`

Commit:
fix tests

Commit:
Merge branch 'hh-exo-deploy6' into hh-exocom-routes

Commit:
Merge branch 'hh-service-env-vars' into hh-terraform-link

Commit:
Merge branch 'hh-terraform-link' into hh-exo-deploy

charlierudolph commented:
Whats the error message look like if you run `exo deploy` without `exo deploy init`?

Commit:
address comments

Commit:
address comments

Commit:
add error message and change command

hugoesthere commented:
@charlierudolph I've added an error message if you run `exo deploy` before running the init command, as well as tests for that. I've also changed `exo deploy init`  to `exo deploy build` since that command can be run more than once and isn't really just an initialization.

Commit:
Merge branch 'hh-exo-deploy6' into hh-exocom-routes

Commit:
Merge branch 'hh-service-env-vars' into hh-terraform-link

Commit:
Merge branch 'hh-terraform-link' into hh-exo-deploy

hugoesthere commented:
Although it seems that this could be causing a race condition. Do you know why that could be?

hugoesthere commented on src/util/exec.go:
```diff
@@ -28,23 +28,23 @@ func Run(dir string, commandWords ...string) (string, error) {
+		return &execplus.CmdPlus{}, errors.Wrapf(err, "Error running '%s'. Output:\n%s", strings.Join(commandWords, " "), cmdPlus.GetOutput())
 	}
-	return nil
+	return cmdPlus, nil
```
I think returning this `cmdPlus` object is causing some issues with race conditions and missing output logs. Is there an easier way to do this? All I want to do is listen for a string and emit an error if that string appears in the output. @charlierudolph

Commit:
revert cmdPlus changes

Commit:
check with os helpers instead

hugoesthere commented:
Nevermind, I went with a different implementation to check that `exo deploy init` has been run. Instead of grabbing the process output I just check that the `main.tf` file exists before running any Terraform commands. Should be good for another look @charlierudolph

Commit:
fix

Commit:
resolve merge conflicts

charlierudolph commented on src/terraform/commands.go:
```diff
@@ -10,6 +10,9 @@ import (
 func RunInit(deployConfig types.DeployConfig) error {
+	if !util.DoesFileExist(deployConfig.TerraformFile) {
+		return errors.New("terraform initialization attempted in directory with no 'main.tf' file. Please make sure 'exo deploy build' has been run")
+	}
```
At this point is running `exo deploy build` ever the wrong thing? If not can we just log something like `Missing 'main.tf' file. Initializing deployment...` and run the deployment initializer?

hugoesthere replied:
The only problem with that is automatically running `exo deploy build` will overwrite the existing `main.tf` file. One of the points of breaking this up into `exo deploy build` and `exo deploy` is that after running the first command, the user can inspect the `main.tf` file and make or add changes, and then deploy with those modifications.

However, overwriting the `main.tf` file is a problem we will have to deal with at some point. I was thinking maybe `exo deploy init` could be passed a flag to tell it whether or not to generate a new `main.tf` file. We could even have it generate a new `main.tf` file just with a different name. @alexdavid thoughts?

charlierudolph replied:
I was expecting this to **never** overwrite `main.tf` and instead only initialize it if not present.

hugoesthere replied:
When the user makes changes to the application then then `exo deploy build` needs to be run again so `main.tf` can be updated, so that deployment will reflect the latest version of the app.

charlierudolph replied:
I meant if you run `exo deploy apply` without running `exo deploy build` then it runs `exo deploy build` for you instead of telling you to run it. If the `main.tf` exists, it just applies it.

hugoesthere replied:
Oh I see, so only when `main.tf` doesn't exist does it run `exo deploy build` automatically, otherwise it's up to the user whether or not they overwrite the existing file.

charlierudolph replied:
Correct. I'm not certain if we want to stop after building if we want the user to be able to inspect the `main.tf` first or just apply it immediately.

hugoesthere replied:
True, I'll stop the process after the build portion finishes.

kevgo commented:
I would prefer that if a command has subcommands, that all possible actions have to be done via subcommands. I.e. change `exo deploy` to `exo deploy run` or something. What do you think?

charlierudolph commented:
I agree with @kevgo. I am not a fan of having `exo deploy` do something and also have subcommands

hugoesthere commented:
Yeah, that felt weird too. I'll make it `exo deploy build` and `exo deploy apply`.

Commit:
split command

charlierudolph commented on src/cmd/shared.go:
```diff
@@ -47,6 +51,46 @@ func getAwsConfig() (types.AwsConfig, error) {
+	}
+	appConfig, err := types.NewAppConfig(appDir)
+	if err != nil {
+		return types.DeployConfig{}, fmt.Errorf("Cannot read application configuration: %s", err)
```
Please use [errors.Wrap](https://godoc.org/github.com/pkg/errors) instead of manually wrapping with `fmt.Errorf`

charlierudolph commented on src/cmd/shared.go:
```diff
@@ -47,6 +51,46 @@ func getAwsConfig() (types.AwsConfig, error) {
+	}
+	homeDir, err := util.GetHomeDirectory()
+	if err != nil {
+		panic(err)
```
Why are we panic-ing instead of returning the error since this function returns an error?

Commit:
run build automatically

Commit:
fix error messages

Commit:
return on build

alexdavid commented on features/help.feature:
```diff
@@ -63,10 +63,15 @@ Feature: help command
+
+      Available Commands:
+        apply       Deploys an Exosphere application to the cloud
+        build       Prepares an Exosphere application to be deployed
```
This is only run the first time deploying right? Maybe this should be called `exo deploy init`?

hugoesthere replied:
Thats what I called it initially, but `build` also pushes your Docker images to ECR and generates your `.tf` files, so it would theoretically be run each time the user makes changes to their app

Commit:
remove unneeded test

hugoesthere commented:
We're going with a different implementation of `exo deploy` as we discussed.
